### ISSUE TITLE
capi-openstack-controller crashes when volume cannot be deleted

### ISSUE THREAD

**Issue by erkanerol** [kubernetes-sigs/cluster-api-provider-openstack#1394]
/kind bug

**What steps did you take and what happened:**

I deleted a machine and then capi-openstack-controller started crashing.

**What did you expect to happen:**

capi-openstack-controller shouldn't crash.

**Anything else you would like to add:**

There was an inconsistent state in `Volume` in OpenStack. It was `in-use` state but the VM was not there. That is why CAPO couldn't delete the VM. It crashed while logging the issue.

I saw this error in OSM's status.
 ```
   failureMessage: 'error deleting OpenStack instance %!s(*compute.InstanceStatus=<nil>)
    with ID Bad request with: [DELETE https://openstack.mydomain:8776/v3/b0c83sdfd94aff9d8c4f1f79990952/volumes/afsdfds32f1-c3aa-5g69-8996-0e454ad8b2ed],
    error message: {"badRequest": {"code": 400, "message": "Invalid volume: Volume
    status must be available or error or error_restoring or error_extending or error_managing
    and must not be migrating, attached, belong to a group, have snapshots or be disassociated
    from snapshots after volume transfer."}}:'
 ```
 
 I saw these errors in CAPO's logs
 ```
 panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x14e7c5e]

goroutine 505 [running]:
sigs.k8s.io/cluster-api-provider-openstack/pkg/cloud/services/compute.(*InstanceStatus).Name(...)
	/workspace/pkg/cloud/services/compute/instance_types.go:88
sigs.k8s.io/cluster-api-provider-openstack/controllers.(*OpenStackMachineReconciler).reconcileDelete(0x1a6c338, {0x1a6c338, 0xc001ada420}, 0xc0016506c0, 0xc0009d0780, 0xc00240ae00, 0xc00214d000, 0xc001c0b440, 0xc0009d0780)
	/workspace/controllers/openstackmachine_controller.go:260 +0x55e
sigs.k8s.io/cluster-api-provider-openstack/controllers.(*OpenStackMachineReconciler).Reconcile(0xc0004bf140, {0x1a6c338, 0xc001ada420}, {{{0xc00073e9c0, 0x1766dc0}, {0xc000701f60, 0x30}}})
	/workspace/controllers/openstackmachine_controller.go:157 +0xb05
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0xc0000c2210, {0x1a6c338, 0xc001ada3f0}, {{{0xc00073e9c0, 0x1766dc0}, {0xc000701f60, 0x413c34}}})
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:114 +0x26f
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc0000c2210, {0x1a6c290, 0xc000659780}, {0x16a4c00, 0xc00043f820})
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:311 +0x33e
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc0000c2210, {0x1a6c290, 0xc000659780})
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:266 +0x205
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:227 +0x85
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:223 +0x357

 ```

The problematic line is here https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/release-0.6/controllers/openstackmachine_controller.go#L260

Somehow `instanceStatus` is nil here.

```
handleUpdateMachineError(scope.Logger, openStackMachine, errors.Errorf("error deleting OpenStack instance %s with ID %s: %v", instanceStatus.Name(), instanceStatus.ID(), err))
		return ctrl.Result{}, nil
```

**Environment:**

- Cluster API Provider OpenStack version (Or `git rev-parse HEAD` if manually built): v0.6.3
- Cluster-API version: v1.1.6
- Kubernetes version (use `kubectl version`): v1.22.8
- OS (e.g. from `/etc/os-release`): Ubuntu 2004

**Comment by jichenjc** [comment]
so , to clarify 

>>>panic: runtime error: invalid memory address or nil pointer dereference

this is something need to be fixed ,definitely
@erkanerol do you want to submit a PR for this? 

>>> status must be available or error or error_restoring or error_extending or error_managing

this is expected, so if we can't delete volume for whatever reason, it's reasonable to keep the volume and instance?

**Comment by erkanerol** [comment]
@jichenjc I haven't checked it in detail. I am questioning which one is correct:
1. `instanceStatus` shouldn't be nil at this line. We should check why it is nil.
2. `instanceStatus ` can be nil at this line. We should fix log line.

**Comment by jichenjc** [comment]
looks like main branch doesn't have this code any more

check this , we should check nil possiblity 
https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/release-0.6/pkg/cloud/services/compute/instance.go#L517

**Comment by mdbooth** [comment]
Yes, this is fixed in main by b9689c61933b5bd5947f6518a1ef345d75645ce9.

Can you get by with a workaround? We're hopefully releasing 0.7 imminently and I'd prefer to avoid another 0.6 release, especially as this would be essentially a custom backport.

To workaround just remove the finalizer from the OpenStackMachine and then delete the OpenStackMachine object. You'll have to remember to sort the errant root volume manually.

**Comment by erkanerol** [comment]
As far as I see, it is enough to remove `instanceStatus` from this line. As a workaround, I can create a fork and build my own version temporarily. 

```
errors.Errorf("error deleting OpenStack instance %s with ID %s: %v", instanceStatus.Name(), instanceStatus.ID(), err)
```
